File: php/features.php
<?php
declare(strict_types=1);

// HubspotAutomation SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class HubspotAutomationFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new HubspotAutomationBaseFeature();
            case "debug":
                return new HubspotAutomationDebugFeature();
            case "idempotency":
                return new HubspotAutomationIdempotencyFeature();
            case "metrics":
                return new HubspotAutomationMetricsFeature();
            case "paging":
                return new HubspotAutomationPagingFeature();
            case "ratelimit":
                return new HubspotAutomationRatelimitFeature();
            case "retry":
                return new HubspotAutomationRetryFeature();
            case "test":
                return new HubspotAutomationTestFeature();
            case "timeout":
                return new HubspotAutomationTimeoutFeature();
            default:
                return new HubspotAutomationBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/hubspotautomation_sdk/features.py
# HubspotAutomation SDK feature factory

from hubspotautomation_sdk.feature.base_feature import HubspotAutomationBaseFeature
from hubspotautomation_sdk.feature.debug_feature import HubspotAutomationDebugFeature
from hubspotautomation_sdk.feature.idempotency_feature import HubspotAutomationIdempotencyFeature
from hubspotautomation_sdk.feature.metrics_feature import HubspotAutomationMetricsFeature
from hubspotautomation_sdk.feature.paging_feature import HubspotAutomationPagingFeature
from hubspotautomation_sdk.feature.ratelimit_feature import HubspotAutomationRatelimitFeature
from hubspotautomation_sdk.feature.retry_feature import HubspotAutomationRetryFeature
from hubspotautomation_sdk.feature.test_feature import HubspotAutomationTestFeature
from hubspotautomation_sdk.feature.timeout_feature import HubspotAutomationTimeoutFeature


_FEATURES = {
    "base": lambda: HubspotAutomationBaseFeature(),
    "debug": lambda: HubspotAutomationDebugFeature(),
    "idempotency": lambda: HubspotAutomationIdempotencyFeature(),
    "metrics": lambda: HubspotAutomationMetricsFeature(),
    "paging": lambda: HubspotAutomationPagingFeature(),
    "ratelimit": lambda: HubspotAutomationRatelimitFeature(),
    "retry": lambda: HubspotAutomationRetryFeature(),
    "test": lambda: HubspotAutomationTestFeature(),
    "timeout": lambda: HubspotAutomationTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
